<a href="https://colab.research.google.com/github/FirmanAhmadBayumi/Machine-Learning/blob/main/PERTEMUAN5_FirmanAhmadBayumi_2117051062.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pertemuan 5

## Library

In [64]:
import pandas as pd
import numpy as np
import sklearn as sk
from sklearn.cluster import AgglomerativeClustering as AC
from sklearn.model_selection import train_test_split as tts
from sklearn.metrics import accuracy_score as accScore
from sklearn.ensemble import RandomForestClassifier as RFS

## Latihan 1

### Data Set

In [ ]:
# membuat data tabular
df = pd.DataFrame(
    {
        "ID Buah" : pd.Series([1, 2, 3, 4, 5]),
        "Diameter" : pd.Series([7.0, 7.1, 6.5, 6.6, 6.8]),
        "Berat": pd.Series([165, 170, 180, 195, 200])
    }
)

df

ID Buah  Diameter  Berat
0        1       7.0    165
1        2       7.1    170
2        3       6.5    180
3        4       6.6    195
4        5       6.8    200

### Preprocessing

In [ ]:
# pemilihan featur
X = df.drop("ID Buah", axis=1)
X

Diameter  Berat
0       7.0    165
1       7.1    170
2       6.5    180
3       6.6    195
4       6.8    200

### Modeling
Menggunakan algoritma agglomerative clustering

In [ ]:
clustering = AC().fit(X)

#hasil clustering
df["kluster"] = clustering.labels_

## Latihan 2

### Dataset

In [ ]:
df2 = pd.read_csv("https://raw.githubusercontent.com/rahmantaufik27/dataset/main/study/student_score_pass.csv")
df2

npm  activeness_score  assignment_score  pre_test_score  \
0    2022041101                99              71.0            93.0   
1    2022041102                96              77.0             NaN   
2    2022041103                86              71.0            84.0   
3    2022041104                81              96.0             NaN   
4    2022041105                93              85.0            68.0   
..          ...               ...               ...             ...   
367  2022041468                91               NaN            31.0   
368  2022041469                96              87.0            55.0   
369  2022041470               569              83.0            89.0   
370  2022041471                84              92.0            87.0   
371  2022041472                90              99.0            96.0   

     post_test_score  midterm_test_score  end_semester_test_score pass_test  
0               52.0                44.0                       24  not pass  
1               27.0               100.0                       89      pass  
2               17.0                21.0                       41  not pass  
3                NaN                11.0                       99  not pass  
4               29.0                44.0                       50  not pass  
..               ...                 ...                      ...       ...  
367             52.0                31.0                       46  not pass  
368             56.0                95.0                       11      pass  
369             15.0                86.0                       13  not pass  
370             97.0                42.0                       27      pass  
371             60.0                82.0                       65  not pass  

[372 rows x 8 columns]

### Preprocessing

In [ ]:
# pemilihan feature
df_training = df2.drop(['npm', 'pass_test'], axis=1)

# hapus NaN values
df_training = df_training.dropna()
df_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  
0                  44.0                       24  
2                  21.0                       41  
4                  44.0                       50  
5                  57.0                       98  
6                  85.0                       13  
..                  ...                      ...  
366                55.0                       82  
368                95.0                       11  
369                86.0                       13  
370                42.0                       27  
371                82.0                       65  

[312 rows x 6 columns]

### Modelling

In [ ]:
model = AC().fit(df_training)

# hasil clustering
df_training["cluster"] = model.labels_
df_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  cluster  
0                  44.0                       24        0  
2                  21.0                       41        0  
4                  44.0                       50        0  
5                  57.0                       98        0  
6                  85.0                       13        0  
..                  ...                      ...      ...  
366                55.0                       82        0  
368                95.0                       11        0  
369                86.0                       13        1  
370                42.0                       27        0  
371                82.0                       65        0  

[312 rows x 7 columns]

## Tugas
Evaluasi model algortima agglomerative clustering dari data latihan 2 dengan mencari nilai akurasinya

Hint: splitting data, modeling, accuracy score

Nilai Akurasi: 0.5079365079365079

In [22]:
# Melakukan pengambilan data

url = "https://raw.githubusercontent.com/rahmantaufik27/dataset/main/study/student_score_pass.csv"
data = pd.read_csv(url)
data

npm  activeness_score  assignment_score  pre_test_score  \
0    2022041101                99              71.0            93.0   
1    2022041102                96              77.0             NaN   
2    2022041103                86              71.0            84.0   
3    2022041104                81              96.0             NaN   
4    2022041105                93              85.0            68.0   
..          ...               ...               ...             ...   
367  2022041468                91               NaN            31.0   
368  2022041469                96              87.0            55.0   
369  2022041470               569              83.0            89.0   
370  2022041471                84              92.0            87.0   
371  2022041472                90              99.0            96.0   

     post_test_score  midterm_test_score  end_semester_test_score pass_test  
0               52.0                44.0                       24  not pass  
1               27.0               100.0                       89      pass  
2               17.0                21.0                       41  not pass  
3                NaN                11.0                       99  not pass  
4               29.0                44.0                       50  not pass  
..               ...                 ...                      ...       ...  
367             52.0                31.0                       46  not pass  
368             56.0                95.0                       11      pass  
369             15.0                86.0                       13  not pass  
370             97.0                42.0                       27      pass  
371             60.0                82.0                       65  not pass  

[372 rows x 8 columns]

### Splitting Data

In [23]:
# Melakukan penghapusan data yang bersifat NaN

data.dropna(inplace = True)
data

npm  activeness_score  assignment_score  pre_test_score  \
0    2022041101                99              71.0            93.0   
2    2022041103                86              71.0            84.0   
4    2022041105                93              85.0            68.0   
5    2022041106                85              95.0            34.0   
6    2022041107                95              79.0            52.0   
..          ...               ...               ...             ...   
366  2022041467                88              75.0            80.0   
368  2022041469                96              87.0            55.0   
369  2022041470               569              83.0            89.0   
370  2022041471                84              92.0            87.0   
371  2022041472                90              99.0            96.0   

     post_test_score  midterm_test_score  end_semester_test_score pass_test  
0               52.0                44.0                       24  not pass  
2               17.0                21.0                       41  not pass  
4               29.0                44.0                       50  not pass  
5               73.0                57.0                       98  not pass  
6               54.0                85.0                       13  not pass  
..               ...                 ...                      ...       ...  
366             11.0                55.0                       82  not pass  
368             56.0                95.0                       11      pass  
369             15.0                86.0                       13  not pass  
370             97.0                42.0                       27      pass  
371             60.0                82.0                       65  not pass  

[312 rows x 8 columns]

In [31]:
# Mengubah data di kolom pass_test not pass menjadi 0, dan pass menjadi 1 
data['pass_test'] = data['pass_test'].replace('not pass', 0)
data['pass_test'] = data['pass_test'].replace('pass', 1)

# Mengkonversi nilai pada kolom pass_test ke tipe data float 
data['pass_test'] = data['pass_test'].astype(float)

### Modelling

In [32]:
# Pemilihan Feature
# Pada program ini saya akan menghapus data pass_test

data_training = data.drop(["npm"], axis = 1)
data_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  pass_test  
0                  44.0                       24        0.0  
2                  21.0                       41        0.0  
4                  44.0                       50        0.0  
5                  57.0                       98        0.0  
6                  85.0                       13        0.0  
..                  ...                      ...        ...  
366                55.0                       82        0.0  
368                95.0                       11        1.0  
369                86.0                       13        0.0  
370                42.0                       27        1.0  
371                82.0                       65        0.0  

[312 rows x 7 columns]

In [58]:
x_dataLatih, x_dataUji, y_dataLatih, y_dataUji = tts(data_training.drop(["pass_test"], axis = 1), data_training["pass_test"], test_size=0.2, random_state=2)

Deskripsi penjelasan pada kode program diatas:
1. test_size digunakan untuk menentukan ukuran data uji sebagai proporsi dari keseluruhan dataset
2. test_size = 0.2, artinya 20% dari data uji yang akan digunakan sebagai data tes, sedangkan 80% sisanya digunakan sebagai data latih.
3. random_state adalah argumen opsional yang digunakan untuk menentukan seed atau biji secara acak yang digunakan dalam pembagian data.



In [59]:
model = AC().fit(x_dataLatih)
model

AgglomerativeClustering()

In [60]:
# Melakukan prediksi pada data latih

predict_data_latih = model.fit_predict(x_dataLatih)

predict_data_latih

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0])

In [61]:
# Melakukan prediksi pada data uji

predict_data_uji = model.fit_predict(x_dataUji)
predict_data_uji

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### Accuracy Score

In [62]:
# Menghitung nilai akurasi pada data latih

hitAkurasi_dataLatih = accScore(y_dataLatih, predict_data_latih)
hitAkurasi_dataLatih

0.642570281124498

In [63]:
# Menghitung nilai akurasi pada data Uji

hitAkurasi_dataUji = accScore(y_dataUji, predict_data_uji)
hitAkurasi_dataUji

0.5873015873015873

### Bonus
Hitung nilai akurasi data lebih dari 0.5

In [84]:
# Random Forest Classifier digunakan dengan jumlah 500 estimators (jumlah pohon keputusan yang akan dibangun) dengan random_state sebesar 3.

model2 = RFS(n_estimators = 500, random_state = 3)
model2

RandomForestClassifier(n_estimators=500, random_state=3)

In [85]:
# Melatih model Random Forest Classifier menggunakan data latih
model2.fit(x_dataLatih, y_dataLatih)

# Melakukan prediksi pada data uji
predict_data_uji2 = model2.predict(x_dataUji)
predict_data_uji2

array([1., 1., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 1., 0.,
       0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 1., 0., 0., 0., 1., 1., 1., 0., 0., 0., 0., 1., 1.,
       0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 0.])

In [86]:
# Menghitung akurasi pada data uji
akurasi = accScore(y_dataUji, predict_data_uji2)

# Menghitung dan mengubah hasil nilai akurasi ke persen
hitAkurasi = akurasi * 100
print("Akurasi :", hitAkurasi, "%")

Akurasi : 79.36507936507937 %


Nilai akurasi yang bisa saya dapatkan dnegan menggunakan Library Random Forest Classifier yaitu 79.36507936507937 %. Pada perhitungan disini saya menggunakan random_state yaitu 3.